client/my-sites/stats/activity-log-day/index.jsx

```diff
@@ -224,18 +234,36 @@ class ActivityLogDay extends Component {
 					onClose={ this.handleCloseDay( hasConfirmDialog ) }
 				>
 					{ hasLogs &&
-						flatMap( rewriteStream( logs ), log => [
-							log.activityId === requestedRestoreActivityId && rewindConfirmDialog,
-							<ActivityLogItem
-								applySiteOffset={ applySiteOffset }
-								disableRestore={ disableRestore }
-								hideRestore={ hideRestore }
-								key={ log.activityId }
-								log={ log }
-								requestRestore={ requestRestore }
-								siteId={ siteId }
-							/>,
-						] ) }
+						map(
+							// Include the Rewind dialog in the right place in the collection
+							flatMap(
+								rewriteStream( logs ),
+								log =>
+									log.activityId === requestedRestoreActivityId
+										? [ rewindConfirmDialog, log ]
+										: [ log ]
+							),
+							( log, index, collection ) =>
+								ActivityLogDay.isRewindDialog( log ) ? (
+									// If this is the Rewind dialog, insert it in timeline
+									log
+								) : (
+									// Otherwise it's a regular event
+									<ActivityLogItem
+										className={ classNames( {
+											// Is this event right before the Rewind dialog?
+											'is-before-dialog': ActivityLogDay.isRewindDialog( collection[ index + 1 ] ),
+										} ) }
+										applySiteOffset={ applySiteOffset }
+										disableRestore={ disableRestore }
+										hideRestore={ hideRestore }
+										key={ log.activityId }
+										log={ log }
+										requestRestore={ requestRestore }
+										siteId={ siteId }
+									/>
+								)
+						) }
```

Comment:
This refactoring is a bit confusing to me: do you think you could try and explain in writing the objective of the code?

It _appears_ as though the goal is to find out which log items occur immediately _before_ a rewind dialog.

The code itself stands out to me for a few reasons:
  - it's unexpected to see `map( flatMap() )` because I think that our goal is do end up with what `flatMap()` is producing, so I'm not sure why `map()` is there (not to say it can't belong there)
  - the use of `collection[ index + 1 ]` is a warning sign because we know that this will end up returning a non-existent element at the end of every spin through the loop - we're intentionally encoding array-out-of-bounds errors
  - the nested transforms are just a bit hard for me to follow

What I think I see happening here is a second loop after an initial loop which is trying to find those items which have a rewind dialog open, predicting from earlier iteration and separating them out. I wonder if we couldn't extract this processing to the data structure before we get into the JSX and end up with one iteration or at least processing of the data which lets us go once through it in React…

Reply:
I'd like to raise the challenge again to think about the data, data structures, and transform operations we are doing _to_ this data so that we can hone in on it in such a way that the logic remains clear and separated from the UI and so that the UI just kind of "falls into place" naturally since our data is prepared for the visual goals.

here's just one thing I'm brainstorming, and I think the assumption it's based on holds - we'll only have one open restore dialog.

> we want to render each item in a list, but if an item has been selected for a restore, then we should display a restore dialog before that item, and the timeline above that restore dialog ought to break

from this specification I suspect we could accomplish this by splitting our array and treating the two halves differently. we could actualize the same specs with a reducer function as well, carrying forward state through its accumulator (to track things like _the last item_ and _the running list of items_) but I think in this case we can get by with a simpler setup which requires no state tracking and no index tracking…

**we have a list of items and up to one of them may have an open restore dialog**

this looks like `_______-____`

we can split this into a _prefix_ list, the possibly open item, and the _postfix_ list. actually, since all we need to do is operate at the first boundary, we can see it as a first half and possibly-empty second half of a list

now we see this with our added `+` dialog - `_____+------` because everything _after_ the rewind dialog is normal

```js
import { findIndex, initial, last } from 'lodash'

const splitByRewind = ( logs, activityId ) => {
	const rewindIndex = findIndex( logs, log => log.activityId === activityId );

	// no dialog is open, just return the front list
	if ( -1 === rewindIndex ) {
		return [ logs, [] ];
	}

	return [
		logs.slice( 0, rewindIndex ), // before the rewind
		logs.slice( rewindIndex ), // at and after rewind
	];
}
```

now down somewhere in our render function we can add in the visual bits…

```js
const [ beforeRewind, afterRewind ] = splitByRewind( rewriteStream( logs ), requestedRestoreActivityId )

…

const LogItem = ( { log, extraClasses } ) => (
	<ActivityLogItem
		className={ extraClasses }
		applySiteOffset={ applySiteOffset }
		disableRestore={ disableRestore }
		hideRestore={ hideRestore }
		key={ log.activityId }
		log={ log }
		requestRestore={ requestRestore }
		siteId={ siteId }
	/>
);

{ /* insert the rewind dialog before an open rewind event and stylize the preceding item */ }
{ initial( beforeRewind ).map( log => <LogItem { ...{ log } } /> }
{ last( beforeRewind ) && <LogItem { ...{ log: last( beforeRewind ), extraClasses: 'is-before-rewind' } } /> }
{ afterRewind.length && rewindConfirmDialog }
{ afterRewind.map( log => <LogItem { ...{ log } } /> }
```

so in some senses we are losing some context by making a local custom `<LogItem />` but what we gain is a more straight-forward declaration of how the UI will look without mixing in the logic of separating and finding the predecessor to the rewind.

could we have done this differently? maybe it would have made it easier still in the render to have returned all of these three distinct places (normal items, the immediate predecessor to the rewind, the items after the rewind dialog) as arrays which would let us take advantage of the safe-defaulting empty state of lists…

```js
// returns [ before, rightBefore, after ]
const splitByRewind = ( logs, activityId ) => {
	const rewindIndex = findIndex( logs, log => log.activityId === activityId );

	// no dialog is open
	if ( -1 === rewindIndex ) {
		return [ logs, [], [], ];
	}

	// dialog is open at first item
	if ( 0 === rewindIndex ) {
		return [ [], [], logs ];
	}

	// first item is immediately before
	if ( 1 === rewindIndex ) {
		return [ [], [ logs[ 0 ] ], logs.slice( 1 ) ];
	}

	// everything else is standard
	return [
		logs.slice( 0, rewindIndex - 1 ),
		[ logs[ rewindIndex - 1 ] ],
		logs.slice( rewindIndex ),
	];
}

…

render() {
	…
	const [ newer, ontop, older ] = splitByRewind( rewriteStream( logs ), requestedRestoreActivityId );
	…
		{ newer.map( log => <LogItem { ...{ log } } /> ) }
		{ ontop.map( log => <LogItem { ...{ log, extraClasses: 'is-before-rewind' } } /> ) }
		{ older && rewindConfirmDialog }
		{ older.map( log => <LogItem { ...{ log } } /> ) }
	…
}
```

hope this gives you some ideas to work with! please feel free to poke me about it